File: GeneticAlgorithm/routines.R
library(testit)
 
knapsack <- function(cap, peso, valor) {
    n <- length(peso)
    pt <- sum(peso) # deben ser enteros en este caso
    assert(n == length(valor))
    vt <- sum(valor) # pueden ser lo que sea
    if (pt < cap) { # cabe todo
        return(vt)
    } else {
        filas <- cap + 1 
        cols <- n + 1 
        tabla <- matrix(rep(-Inf, filas * cols),
                       nrow = filas, ncol = cols)
        for (fila in 1:filas) {
            tabla[fila, 1] <- 0 
        }
        rownames(tabla) <- 0:cap 
        colnames(tabla) <- c(0, valor) 
        for (objeto in 1:n) { 
            for (acum in 1:(cap+1)) { 
                anterior <- acum - peso[objeto]
                tabla[acum, objeto + 1] <- tabla[acum, objeto]                
                if (anterior > 0) { 
                    tabla[acum, objeto + 1] <- max(tabla[acum, objeto], tabla[anterior, objeto] + valor[objeto])
                }
            }
        }
        return(max(tabla))
    }
}
 
factible <- function(seleccion, pesos, capacidad) {
    return(sum(seleccion * pesos) <= capacidad)
}
 
objetivo <- function(seleccion, valores) {
    return(sum(seleccion * valores))
}
 
al.azar <- function(cuantos) {
    return(round(runif(cuantos)))
}
 
normalizar <- function(data) {
    menor <- min(data)
    mayor <- max(data)
    rango <- mayor - menor
    data <- data - menor # > 0
    return(data / rango) # entre 0 y 1
}
 
generador.pesos <- function(cuantos, min, max) {
    return(sort(round(normalizar(rnorm(cuantos)) * (max - min) + min)))
}
 
generador.valores <- function(pesos, min, max) {
    n <- length(pesos)
    valores <- double()
    for (i in 1:n) {
        media <- pesos[i]
        desv <- runif(1)
        valores <- c(valores, rnorm(1, media, desv))
    }
    valores <- normalizar(valores) * (max - min) + min
    return(valores)
}
 
n <- 20
pesos <- generador.pesos(n, 15, 80)
valores <- generador.valores(pesos, 10, 500)
capacidad <- round(sum(pesos) * 0.65)
print(pesos)
print(valores)
print(knapsack(capacidad, pesos, valores))
for (i in 1:10) {
    intento <- al.azar(n)
    print(paste(factible(intento, pesos, capacidad), objetivo(intento, valores)))
}

File: GeneticAlgorithm/routines.py
import numpy as np
 
def knapsack(peso_permitido, pesos, valores):
    assert len(pesos) == len(valores)
    peso_total = sum(pesos)
    valor_total = sum(valores)
    if peso_total < peso_permitido: 
        return valor_total
    else:
        V = dict()
        for w in range(peso_permitido + 1):
            V[(w, 0)] = 0
        for i in range(len(pesos)):
            peso = pesos[i]
            valor = valores[i]
            for w in range(peso_permitido + 1):
                cand = V.get((w - peso, i), -float('inf')) + valor
                V[(w, i + 1)] = max(V[(w, i)], cand)
        return max(V.values())
 
def factible(seleccion, pesos, capacidad):
    return np.inner(seleccion, pesos) <= capacidad
  
def objetivo(seleccion, valores):
    return np.inner(seleccion, valores)
 
def al_azar(cuantos):
    return np.round(np.random.uniform(size = cuantos))
  
def normalizar(data):
    menor = min(data)
    mayor = max(data)
    rango  = mayor - menor
    data = data - menor # > 0
    return data / rango # entre 0 y 1
  
def generador_pesos(cuantos, low, high):
    return np.round(normalizar(np.random.normal(size = cuantos)) * (high - low) + low)
 
from random import random
def generador_valores(pesos, low, high):
    n = len(pesos)
    valores = np.empty((n))
    for i in range(n):
        valores[i] = np.random.normal(pesos[i], random(), 1)
    return normalizar(valores) * (high - low) + low
 
n = 20
pesos = generador_pesos(n, 15, 80)
valores = generador_valores(pesos, 10, 500)
capacidad = int(round(sum(pesos) * 0.65))
print(pesos)
print(valores)
print(knapsack(capacidad, pesos, valores))

def replica(rID):
    intento = al_azar(n)
    return (factible(intento, pesos, capacidad), objetivo(intento, valores)) 

import multiprocessing as mp
if __name__ == '__main__':
    pool = mp.Pool(mp.cpu_count() - 1)
    salida = pool.map(replica, range(10))
    for s in salida:
        print(s)
